net/FlatBuffers/ByteBuffer.cs

```diff
@@ -40,21 +40,25 @@ public class ByteBuffer
     {
         private readonly byte[] _buffer;
         private int _pos;  // Must track start of the buffer.
+        private int _off; // parent's position if sliced
 
         public int Length { get { return _buffer.Length; } }
 
         public byte[] Data { get { return _buffer; } }
 
         public ByteBuffer(byte[] buffer) : this(buffer, 0) { }
 
-        public ByteBuffer(byte[] buffer, int pos)
+        public ByteBuffer(byte[] buffer, int pos) : this(buffer, pos, 0) { }
+
+        private ByteBuffer(byte[] buffer, int pos, int off)
         {
             _buffer = buffer;
             _pos = pos;
+            _off = off;
         }
 
         public int Position {
-            get { return _pos; }
+            get { return _pos + _off; }
```

Comment:
We didn't need this `_off` in Java, why here? It is both not great for performance (since `Position` is used a lot), and makes it harder to tell if this code is correct, i.e. if `_off` is taken into account in all the right places.

Reply:
There is probably a way to implement this a little more efficiently using `unsafe` C# if that does not violate your guidelines. Note that the Java `ByteBuffer` implementation does something similar for each `get`, e.g. in `HeapByteBuffer`:

```Java
protected int ix(int i) {
    return i + offset;
}

public byte get(int i) {
    return hb[ix(checkIndex(i))];
}

final int checkIndex(int i) {
    if ((i < 0) || (i >= limit))
        throw new IndexOutOfBoundsException();
    return i;
}
```

Reply:
My final note on this will be that I'll vote for a similar abstraction in C# as in Java:

```csharp
using System;

namespace FlatBuffers
{
    public class ByteBuffer
    {
        private readonly byte[] _buffer;
        private readonly int _cap; // calculated capacity
        private int _pos;          // Must track start of the buffer.
        private readonly int _off; // parent's position if sliced

        public int Length { get { return _cap; } }

        public byte Get(int pos) { return _buffer[_off + pos]; }

        public void Put(int pos, byte b) { _buffer[_off + pos] = b; }

        public ByteBuffer(byte[] buffer) : this(buffer, 0) { }

        public ByteBuffer(byte[] buffer, int pos) : this(buffer, pos, 0) { }

        private ByteBuffer(byte[] buffer, int pos, int off)
        {
            _buffer = buffer;
            _pos = pos;
            _off = off;
            _cap = buffer.Length - _off;
        }

        public int Position {
            get { return _pos; }
            set { _pos = value; }
        }

        public void Reset()
        {
            _pos = 0;
        }

        // Create a new ByteBuffer on the same underlying data.
        // The new ByteBuffer's position will be 0, but starting
        // from this ByteBuffer's current position.
        public ByteBuffer Slice()
        {
            return new ByteBuffer(_buffer, 0, Position);
        }

        // ...
}
```